crates/schema-inference/src/server.rs

```diff
@@ -0,0 +1,224 @@
+            match shape {
+                Ok(Some((inferred_shape, docs_read))) => {
+                    if let Some((accumulated_shape, docs_count)) = accumulator {
+                        accumulator = Some((shape::merge(accumulated_shape, inferred_shape), docs_count + docs_read))
+                    } else {
+                        accumulator = Some((inferred_shape, docs_read))
+                    }
+                }
+                Ok(None) => {}
+                Err(e) => return Err(e),
+            }
+        }
+
+        match accumulator {
+            Some((shape, docs_count)) => {
+                let root_schema = SchemaBuilder::new(shape).root_schema();
+                let root = serde_json::ser::to_string(&root_schema)
+                    .map_err(|e| Status::internal(e.to_string()))?;
+                Ok(Response::new(InferenceResponse {
+                    body: Some(schema_inference::inference_response::Body::Schema(
+                        schema_inference::InferredSchema {schema_json:root, documents_read: docs_count },
+                    )),
+                }))
+            }
+            None => Err(Status::not_found("No documents found, cannot infer shape")),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct JournalMeta {
+    uuid: String,
+    #[serde(skip_serializing)]
+    ack: Option<bool>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct JournalMessage {
+    #[serde(rename = "_meta")]
+    meta: JournalMeta,
+    #[serde(flatten)]
+    body: Map<String, Value>,
+}
+
+async fn journal_to_shape(
+    journal: JournalSpec,
+    client: journal_client::Client,
+) -> Result<Option<(Shape, u64)>, Status> {
+    // TODO: Split this request up into parallelizable chunks.
+    // The reason we can't just split in to N even-sized chunks is that
+    // we could very easily split in the middle of a document.
+    // One option is to enumerate fragments in order to determine reasonable
+    // split points, and then use the Reader to read to those in parallel.
+
+    let reader = Reader::start_read(
+        client.clone(),
+        JournalRead::new(journal.name.clone()),
+        NoRetry,
+    );
+
+    let codec = LinesCodec::new(); // do we want to limit length here? LinesCodec::new_with_max_length(...) does this
+    let mut doc_bytes_stream = FramedRead::new(FuturesAsyncReadCompatExt::compat(reader), codec);
+
+    let mut accumulator: Option<Shape> = None;
+    let mut docs: u64 = 0;
+
+    while let Some(maybe_doc_body) = doc_bytes_stream.next().await {
+        match maybe_doc_body {
+            Ok(doc_str) => {
+                let parsed: JournalMessage = serde_json::de::from_str(doc_str.as_str())
+                    .map_err(|e| Status::aborted(e.to_string()))?;
+                // There should probably be a higher-level API for this in `journal-client`
+
+                if parsed.meta.ack.is_none() {
+                    let re_serialized = to_value(parsed).unwrap();
```

Comment:
This re-serialization could be mitigated by getting rid of the `JournalMessage` and `JournalMeta` structs entirely and just deserializing directly to a `serde_json::Value`. You can use the `Value::pointer` function to check whether `/_meta/ack` is true.